src/lightgbm/src/main/scala/LightGBMUtils.scala

```diff
@@ -0,0 +1,233 @@
+    (lightgbmlib.double_to_voidp_ptr(data), data)
+  }
+
+  def generateDenseDataset(numRows: Int, rowsAsDoubleArray: Array[Array[Double]]): SWIGTYPE_p_void = {
+    val numRowsIntPtr = lightgbmlib.new_intp()
+    lightgbmlib.intp_assign(numRowsIntPtr, numRows)
+    val numRows_int32_tPtr = lightgbmlib.int_to_int32_t_ptr(numRowsIntPtr)
+    val numCols = rowsAsDoubleArray.head.length
+    val isRowMajor = 1
+    val numColsIntPtr = lightgbmlib.new_intp()
+    lightgbmlib.intp_assign(numColsIntPtr, numCols)
+    val numCols_int32_tPtr = lightgbmlib.int_to_int32_t_ptr(numColsIntPtr)
+    val datasetOutPtr = lightgbmlib.voidpp_handle()
+    val datasetParams = "max_bin=255 is_pre_partition=True"
+    val data64bitType = lightgbmlibConstants.C_API_DTYPE_FLOAT64
+    var data: Option[(SWIGTYPE_p_void, SWIGTYPE_p_double)] = None
+    try {
+      data = Some(generateData(numRows, rowsAsDoubleArray))
+      // Generate the dataset for features
+      LightGBMUtils.validate(lightgbmlib.LGBM_DatasetCreateFromMat(data.get._1, data64bitType,
+        numRows_int32_tPtr, numCols_int32_tPtr, isRowMajor, datasetParams, null, datasetOutPtr), "Dataset create")
+    } finally {
+      if (data.isDefined) {
+        lightgbmlib.delete_doubleArray(data.get._2)
+      }
+    }
+    lightgbmlib.voidpp_value(datasetOutPtr)
+  }
+
+  /** Generates a sparse dataset in CSR format.
+    * @param sparseRows The rows of sparse vector.
+    * @return
+    */
+  def generateSparseDataset(sparseRows: Array[SparseVector]): SWIGTYPE_p_void = {
+    var values: Option[(SWIGTYPE_p_void, SWIGTYPE_p_double)] = None
+    var indexes: Option[(SWIGTYPE_p_int32_t, SWIGTYPE_p_int)] = None
+    var indptrNative: Option[(SWIGTYPE_p_int32_t, SWIGTYPE_p_int)] = None
+    try {
+      val valuesArray = sparseRows.flatMap(_.values)
+      values = Some(newDoubleArray(valuesArray))
+      val indexesArray = sparseRows.flatMap(_.indices)
+      indexes = Some(newIntArray(indexesArray))
+      val indptr = new Array[Int](sparseRows.length + 1)
+      sparseRows.zipWithIndex.foreach {
+        case (row, index) => indptr(index + 1) = indptr(index) + row.numNonzeros
+      }
+      indptrNative = Some(newIntArray(indptr))
+      val numCols = sparseRows(0).size
+
+      val datasetOutPtr = lightgbmlib.voidpp_handle()
+      val datasetParams = "max_bin=255 is_pre_partition=True"
+      val dataInt32bitType = lightgbmlibConstants.C_API_DTYPE_INT32
+      val data64bitType = lightgbmlibConstants.C_API_DTYPE_FLOAT64
+
+      // Generate the dataset for features
+      LightGBMUtils.validate(CSRUtils.LGBM_DatasetCreateFromCSR(indptrNative.get._1, dataInt32bitType,
+        indexes.get._1, values.get._1, data64bitType, intToPtr(indptr.length), intToPtr(valuesArray.length),
+        intToPtr(numCols), datasetParams, null, datasetOutPtr), "Dataset create")
+      lightgbmlib.voidpp_value(datasetOutPtr)
+    } finally {
+      // Delete the input rows
+      if (values.isDefined) {
```

Comment:
Pretty sure delete_doubleArray would be a no-op if the pointer is null but I guess it doesn't hurt to check..